sysutils/apcupsd/src/opnsense/service/conf/actions.d/actions_apcupsd.conf

```diff
@@ -0,0 +1,29 @@
+[start]
+command:/usr/sbin/service apcupsd start && sleep 2
```

Comment:
sleep is necessary? usually we use rc.d directly to avoid bouncing through "service" utility
```suggestion
command:/usr/local/etc/rc.d/namedapcupsd start && sleep 2
```

Reply:
Will check with @Gibbon99 about the sleep.  But if the /usr/local/etc/rc.d/ works ok then by all means. :-)
//Dan Lundqvist (mrzaz)

Reply:
I added the sleep because the service was not showing as started when requesting its status after.

For example when enabling the service and applying the configuration it was showing as stopped.

That's also why I implemented `waitEnabled` function here https://github.com/xbb/opnsense-plugins/blob/851a320b4fc7a8100948a18176adb7def923d488/sysutils/apcupsd/src/opnsense/mvc/app/views/OPNsense/Apcupsd/index.volt#L45-L73 which retries to update the service control UI for max 5 times (1 sec between request) until the service is detected as running, just in case if it takes more than 2 seconds to start

test case:

```sh
/usr/local/etc/rc.d/apcupsd onestart; /usr/local/etc/rc.d/apcupsd onestatus
```

Do you know if there is a better way to handle this?

I could wrap `/usr/local/etc/rc.d/apcupsd start` with a shell script and wait for the pidfile `/var/run/apcupsd.pid` to be created with a timeout

Reply:
Something like this:

```sh
#!/usr/bin/env sh

set -e

/usr/local/etc/rc.d/apcupsd "$@"

case "$1" in
  start|onestart|restart|onerestart)
    i=0
    while [ ! -f /var/run/apcupsd.pid ] && [ "$i" -le 5 ]; do
      i=$(( i + 1 ))
      sleep 0.5
    done
  ;;
esac
```

Reply:
the GUI poll is nicer than a sleep underneath if this seems to be a general issue with apcupsd... it bears risk of clogging up the backend calls. However, older services sometimes rely on a pgrep "servicename" to get the status since some form of daemon should be running when the pid file isn't there.

Reply:
I just found out I can use a post start/restart script, using the existing template for starting the service

```sh
{% if helpers.exists('OPNsense.apcupsd.general.Enabled') and OPNsense.apcupsd.general.Enabled == "1" %}
apcupsd_enable="YES"
{% else %}
apcupsd_enable="NO"
{% endif %}

# Hook to post start/restart commands waiting for the pid file for max 3 seconds
# this prevents the status command to erroneously report
# the service as stopped since apcupsd detaches before the pid file creation

start_postcmd="${name}_post_start_restart"
restart_postcmd="${name}_post_start_restart"

apcupsd_post_start_restart()
{
    if [ -n "$pidfile" ]; then
        i=0
        while [ ! -f "$pidfile" ] && [ "$i" -le 5 ]; do
          i=$(( i + 1 ))
          sleep 0.5
        done
    fi
}
```

this seems to work well.

The problem is while the JS works well enough for the settings page, it's still a problem in the services widget and services pages where if you start or restart the service it will just show as stopped until the page is reloaded.

![services-restart](https://user-images.githubusercontent.com/239074/152659598-2cdb534e-5bc1-469d-a93f-41b7cd6d3cda.gif)